### DISCUSSION TITLE
Pybricks - Multitask - moving Sensor & scanning Distance

### DISCUSSION THREAD

**Discussion by Slaiderno46** [pybricks/support#1891] [Q&A]
Hello Pybricks´ users,

I´ve made a moving robot from LEGO BOOST (move HUB) with turning head back and forth in some range & distance sensor placed on the head.

I can´t figure out how to code the robot to simultaneously moving head back and forth and scanning the distance from an obstacles.

Now, the robot moves the head from one end position to the other and then scans the distance. This is not how it is supposed to work. I want the sensor to scan the distance several times during range of motion.

I attach the code as I have it now:

Photo of the CODE:
![pybricks_program_2024-10-21T16_23_52](https://github.com/user-attachments/assets/5bba397f-53f9-4207-9ab3-12015600990c)

Photo of the ROBOT:
![IMG_20241021_183555](https://github.com/user-attachments/assets/4933cf98-642c-46e0-9e6f-62a1551130ce)

Transcript of the CODE:

```python
from pybricks.hubs import MoveHub
from pybricks.parameters import Color, Direction, Port, Stop
from pybricks.pupdevices import ColorDistanceSensor, Motor
from pybricks.robotics import DriveBase
from pybricks.tools import multitask, run_task, wait

# Set up all devices.
Boost_Hub = MoveHub()
sensor = ColorDistanceSensor(Port.D)
Right_Side = Motor(Port.A, Direction.COUNTERCLOCKWISE)
Left_Side = Motor(Port.B, Direction.CLOCKWISE)
motor = Motor(Port.C, Direction.COUNTERCLOCKWISE)
drive_base = DriveBase(Left_Side, Right_Side, 56, 114)

async def subtask():
    await motor.run_until_stalled(200, Stop.COAST, 40)
    print('Rotating head to one side')
    await wait(50)
    await motor.run_until_stalled(-200, Stop.COAST, 40)
    print('Rotating head to another side')
    await wait(50)

async def subtask2():
    if await sensor.distance() > 90:
        await sensor.light.on(Color.BLUE)
        print('Light Blue')
        await wait(20)
        Left_Side.run(200)
        Right_Side.run(200)
        print('Keep going')
    else:
        await sensor.light.on(Color.RED)
        print('Light Red')
        await wait(20)
        Left_Side.stop()
        Right_Side.stop()
        print('Stop')
        await drive_base.turn(10, Stop.COAST)
        print('Turn')

async def main():
    print('Hello, Boost! Let´s start!')
    while True:
        await wait(1)
        print('REPEAT ROUND')
        await multitask(subtask(), subtask2())

run_task(main())
```

Thank you very much for help me figure it out. Thanks.

**Comment by laurensvalk** [comment]
I think this is a structure you could use. I've omitted the actual blue movement blocks but added comments where each part would go:

![image](https://github.com/user-attachments/assets/fb6617ae-c83f-46e1-b438-aa897ce0bf90)

**Reply by Slaiderno46** [reply]
Dear Lauren, this is wonderful..! This is not the way I was thinking..! Hopefully I will get better..

Thank you so much..! I´ll try and let you know..

**Reply by Slaiderno46** [reply]
> I think this is a structure you could use. I've omitted the actual blue movement blocks but added comments where each part would go:
> 

Lauren, it works perfectly. Thank you very much for your valuable advice..!

Please, let me have one more ask.. there is a printscreen from PyBricks Help Manual:

![image](https://github.com/user-attachments/assets/3169cf13-38c6-4af9-8582-30270a1da729)

I can´t get it into PyBricks, in other words, I don´t know at some points what icons to be used.

Specifically this part:

left_end = example_motor.run_until_stalled(-200, duty_limit=30)
right_end = example_motor.run_until_stalled(200, duty_limit=30)

example_motor.reset_angle((right_end - left_end) / 2)

How can I save the actual angle into parameter left_end or right_end .. ?
How can I then put into PyBricks (right_end - left_end) / 2 .. ?

Thank you very much in advance..

**Reply by laurensvalk** [reply]
Here you go!

![image](https://github.com/user-attachments/assets/b72c4e3e-aedb-4c29-813e-131f17c13568)

**Reply by Slaiderno46** [reply]
Superb..!!! Now, I see that.. Fantastic..!!!

**Reply by Slaiderno46** [reply]
> Here you go!
> 
I could see that you described coding (basic button-sensing techniques) for control via LEGO Powered Up Remote Control.

Is it possible to let the robot or vehicle drive autonomously and at some time to press a button, start driving via controller and then press a button and go back into autonomous mode..?

I see an issue that LEGO Powered Up Remote Control have just 4 buttons which all you need to control movement.

So, is it possible to press for example two buttons to switch between the modes..?

Thank you.

**Reply by laurensvalk** [reply]
> is it possible

Put another way, can you think of a way in words to describe what should happen?

Perhaps similar to the earlier answers you could do:

- multi task until one of these two tasks complete:
     - drive autonomously
     - wait until remote center button released and then pressed
- multi task until one of these two tasks complete:
     - drive around using the remote
     - wait until remote center button released and then pressed

(hint: the answer is yes :smile:)

**Reply by Slaiderno46** [reply]
That´s insane..! It would never come to my mind to think this way..! Simply beautiful...!

**Reply by Slaiderno46** [reply]
> Here you go!
> 

Dear Lauren, I coded it as below:

![pybricks_program_2024-10-23T16_41_27](https://github.com/user-attachments/assets/a5ef23c7-bade-427d-8268-ae3a37e4d194)

The head is going to one end point, to another end point. Both angles are read. But then the program gives an error:

![image](https://github.com/user-attachments/assets/45725bd7-9363-4b62-8c34-b56831f7461b)

The code fails at command "motor.reset_angle((right_end - left_end) / 2)". It seems like the program would read wrong data which are under parameters right_end and left_end.

**Reply by Slaiderno46** [reply]
There is the issue..:

![image](https://github.com/user-attachments/assets/9095e5b9-1dde-4e05-a4bb-5127613ff244)

**Reply by Slaiderno46** [reply]
With this the code works (math operation "minus"):

![image](https://github.com/user-attachments/assets/d4a0ee2b-0bfe-47df-b00b-2483d46659e6)

With this the code doesn´t work (math operation "divide"):

![image](https://github.com/user-attachments/assets/562bbed5-f65b-422f-a4bf-324c36117122)

Wow, I´ve figured it out.. :) There is an issue in PyBricks..

**Reply by laurensvalk** [reply]
Yes, this is a limitation of the boost move hub. 

How about this: instead of resting it to left and right, just reset it one way (only left) and call that zero. No math required. Then you can still do run towards target, just with a number of degrees of your choosing.

**Reply by Slaiderno46** [reply]
First of all, sorry I called you Lauren. I remember football player from Arsenal named Lauren. And I think many French people are Lauren. Sorry for that.

Yes, you are right. This is a solution - not general, but for specific set-up. Nothing can be done. I think it´s time for MindStorm.. :)

**Reply by laurensvalk** [reply]
No worries at all :smile:

Yes, the Boost Move Hub has some limitations, but it's still a fun one for many builds!

**Comment by BertLindeman** [comment]
Hello,

The movehub does not support floating point arithmetic.
The divide produces a float. I assume **that** is the problem.

IIRC there is an issue asking for the divide in the movehub to be an integer divide (if that is the correct term)

No support now for this divide in the blocks.
On the python code (I copied) I changed the divide into 
```python

center = left_end + left_end // 2

```
And that does not suffer the error.

By the way (laurensvalk has as first name Laurens) 
Had to note it 😄 

Bert

**Reply by Slaiderno46** [reply]
I tried put there number to get result as whole number.. specifically 360/2 but still doesn´t work.

Just for info. The MOVE HUB can´t handle DIVIDING operation..

**Reply by dlech** [reply]
It can do integer division. Just need two slashes instead of one. `//` instead of `/`.

**Reply by Slaiderno46** [reply]
Then I would need PyBricks to be able to ROUND the number to get INTEGER and I don´t know how to edit the Pybricks code from / to //.

**Reply by laurensvalk** [reply]
This will come to a future version of block coding. See https://github.com/pybricks/support/issues/1503

**Reply by Slaiderno46** [reply]
I promise that this is last question, then the robot is finished.. :)

I´ve already leant bunch of things on it.

Now, I wanted to code Powered Up Remote to control it manually. And it works perfectly.

The code is below:

![pybricks_program_2024-10-24T16_30_23](https://github.com/user-attachments/assets/6769ffba-9067-44ae-aa84-1d6344c175c9)

It can go straight and back. It can turn in place.

Now can´t figure out how to combine two buttons to curve. -> when I push LEFT button I want robot to turn in place X when I push LEFT + FORWARD buttons I want robot to curve.

I thought that the easiest way to handle it is to use logic AND function.

![image](https://github.com/user-attachments/assets/fc9ddaeb-2fa9-4cea-95b2-d3bd5b84c4dd)

But it doesn´t work.. I apologize to bother everyone..

**Reply by dlech** [reply]
There is already a check for "if remote left + is pressed" before the check for "if remote right + is pressed and if remote left + is pressed". So the first branch will be taken and the second test never reached. You can fix it by moving the "and" test before the others.

**Reply by Slaiderno46** [reply]
> There is already a check for "if remote left + is pressed" before the check for "if remote right + is pressed and if remote left + is pressed". So the first branch will be taken and the second test never reached. You can fix it by moving the "and" test before the others.

You´re right. It works. That´s really great. Thank you so much. Robot is complete.. :)

Just the Move HUB from Lego Boost doesn´t allow to configure acceleration or maximum speed. It gives error. What a shame.

Thank you very much..!

**Reply by Slaiderno46** [reply]
A little issue. I put both modes (manual via controller and autonomous) into together.

The code works a) from the very beginning robot is in manual mode and it works b) when I push central button the robot switches into autonomous mode and there it works as well.

But the issue occurs when I push the central button again to get back into manual mode.

I don´t understand why when it works firstly from manual to autonomous.

![image](https://github.com/user-attachments/assets/a2582e2d-0101-4271-b15d-8985eba3be9a)

Really sorry to bother everyone. I appreciate your help so much..!

**Reply by laurensvalk** [reply]
Both of the loops run until the button is pressed.

But you don't wait for a release. So it is immediately going around the outer loop again.


So _after_ both versions of the "repeat until remote center pressed" loops, add: wait until not button pressed

**Reply by Slaiderno46** [reply]
Ahh, now I understand it. The robot take the button like being pushed forever so there is no possibility to detect another "push".

So, I revised it as you suggested:

![image](https://github.com/user-attachments/assets/1526a621-4c87-4409-8b00-0ff25b68dbd8)

But it still does NOT work. It´s still an issue to detect button and go back from autonomous mode to manual one.

**Reply by laurensvalk** [reply]
It would be hard to say without seeing what's in the the two drive blocks you made.

The way it is set up now, it will only check for a button press each time your drive function completes. If something inside your drive function takes some time, like waiting for sensors, etc, then it will almost never be checking for the button.

This is where that multitasking suggestion above could come in helpful.

**Reply by Slaiderno46** [reply]
Laurens, you were completely on spot..!

It works.. I changed one of the two loop to "Multitask Until". For each task different loop is suitable.

Nothing more to be solved.. :)

![pybricks_program_2024-10-25T13_03_00](https://github.com/user-attachments/assets/abd497d1-81ed-4f00-a5b1-81a399761199)

**Reply by laurensvalk** [reply]
Excellent! Glad to hear you've enjoyed working with Pybricks :smile:

**Comment by Slaiderno46** [comment]
It was more than that.

Now, I must approach some simple tasks in PyBricks with my 10 years old son.

If you knew about appropriate web site with tasks to learn coding, please
let me know.

I've learned so much things thanks to you.

Once more, thank you Laurens..!

Dne pá 25. 10. 2024 15:56 uživatel laurensvalk ***@***.***>
napsal:

**Reply by BertLindeman** [reply]
> If you knew about appropriate web site with tasks to learn coding, please
let me know.

If you read around the advertisements a very basic start is in [learn python for kids/](https://teachyourkidscode.com/learn-python-for-kids/)
But there are also  books you might find on the web or in the local library
I do not have experience with them, so I advise to do a search.

Bert
